{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c21a51a",
   "metadata": {},
   "source": [
    "# Collective finite responses ORA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee0eb629",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from json import dumps\n",
    "import logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "import json\n",
    "from tqdm.notebook import tqdm\n",
    "import copy\n",
    "\n",
    "from scipy.stats import hypergeom as hg\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "\n",
    "#from scipy.stats import hypergeom as hg\n",
    "import statsmodels.stats as sts\n",
    "\n",
    "from CoRe import reader, fnGO\n",
    "#from multipy.fdr import qvalue\n",
    "\n",
    "import random as rand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "56b07245",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_set(dropout_dict):\n",
    "    selected_set = []\n",
    "    \n",
    "    for gene in dropout_dict.keys():\n",
    "        xi = rand.uniform(1,0)\n",
    "        \n",
    "        if xi>dropout_dict[gene]:\n",
    "            selected_set.append(gene)\n",
    "            \n",
    "    selected_set = pd.DataFrame(selected_set,columns=None)\n",
    "            \n",
    "    return selected_set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5fa640d",
   "metadata": {},
   "source": [
    "**Read the gene sets and determine the list of unique genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d986fda2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total gene sets:  7481\n",
      "Total unique genes:  17949\n"
     ]
    }
   ],
   "source": [
    "go_directory = \"./Examples/GO_sets\"\n",
    "os.chdir(go_directory)\n",
    "\n",
    "f = open(('GO_BPs.json'))\n",
    "GO_BPs = json.load(f)\n",
    "f.close()\n",
    "\n",
    "for bp in GO_BPs.keys():\n",
    "    GO_BPs[bp] = pd.DataFrame(GO_BPs[bp])\n",
    "\n",
    "total_gene_set = pd.read_csv('all_unique_genes.csv',header=None)[0].to_list()\n",
    "total_genes = len(total_gene_set)\n",
    "\n",
    "print('Total gene sets: ',len(list(GO_BPs.keys())))\n",
    "print('Total unique genes: ',total_genes)\n",
    "\n",
    "wf = open('BP_embedding.csv')\n",
    "all_lines = wf.readlines()\n",
    "wf.close()\n",
    "\n",
    "embed_idx = {}\n",
    "\n",
    "for l in all_lines:\n",
    "    all_values = l.rstrip('\\r\\n').split(',')\n",
    "\n",
    "    embed_idx[int(all_values[0])] = []\n",
    "\n",
    "    for k in all_values[1:]:\n",
    "        embed_idx[int(all_values[0])].append(int(k))\n",
    "        \n",
    "bp_names = pd.read_csv('GOBP_list.csv',header=None)[0].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b7a16f",
   "metadata": {},
   "source": [
    "Read the set of genes that have direct protein-protein interactions with SARS-CoV-2 proteins, and read the set of gene that are both directly interacting or indirectly receiving information about the SARS-CoV-2 proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7c04c0c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_pathway = 'Immune System'\n",
    "pathway_nametag = selected_pathway.replace(' ','_')\n",
    "\n",
    "network_type = 'medium-PPI'\n",
    "network_label = 'medium'\n",
    "state_type = 'maxEnt'\n",
    "\n",
    "data_directory = \"/Users/swarnavo/CodeX/CoRe/Examples/\"+pathway_nametag\n",
    "os.chdir(data_directory)\n",
    "\n",
    "direct_interaction_set = reader.read_interactions_for_GO('SARS_CoV2-'+pathway_nametag+'_interactions.json')\n",
    "communicated_genes = pd.read_csv('collectively_communicated_proteins.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "86aabe43",
   "metadata": {},
   "outputs": [],
   "source": [
    "directly_affected_genes = []\n",
    "\n",
    "for n in list(direct_interaction_set):\n",
    "    directly_affected_genes += direct_interaction_set[n][0].to_list()\n",
    "    \n",
    "directly_affected_genes = list(set(directly_affected_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0742ef99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'collective_action':            0\n",
      "0     IMPDH2\n",
      "1        PVR\n",
      "2        GLA\n",
      "3      RAB7A\n",
      "4       TBK1\n",
      "...      ...\n",
      "1043  BTN3A3\n",
      "1044  BTN3A2\n",
      "1045   BTNL9\n",
      "1046  BTN2A1\n",
      "1047   BTNL2\n",
      "\n",
      "[1048 rows x 1 columns]}\n"
     ]
    }
   ],
   "source": [
    "data = {}\n",
    "\n",
    "data[0] = directly_affected_genes + communicated_genes['Gene'].to_list()\n",
    "\n",
    "total_affected_genes = {}\n",
    "total_affected_genes['collective_action'] = pd.DataFrame.from_dict(data)\n",
    "\n",
    "print(total_affected_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "910c4ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_tags, go_names, p_values = fnGO.compute_p_values(['collective_action'],GO_BPs,total_affected_genes,total_genes)#,size_threshold=gene_set_cutoff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "796192a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_tags, go_names, q_values = fnGO.compute_q_values(p_values['collective_action'],go_names['collective_action'],go_tags['collective_action'],alpha=0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cd3afda7",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_idx = []\n",
    "\n",
    "for go in go_tags:\n",
    "    go_idx.append(bp_names.index(go))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6d20a966",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx_to_remove = []\n",
    "    \n",
    "k = 0\n",
    "\n",
    "for ii in go_idx:\n",
    "    idx_copy = copy.deepcopy(go_idx)\n",
    "    idx_copy.remove(ii)\n",
    "\n",
    "    d1 = pd.DataFrame(embed_idx[ii])\n",
    "    d2 = pd.DataFrame(idx_copy)\n",
    "\n",
    "    if len(d1)>0 and len(d2)>0:\n",
    "        intersection = pd.merge(d1, d2, how='inner').drop_duplicates([0])\n",
    "        len_intersection = int(intersection[0].count())\n",
    "\n",
    "        if len_intersection>0:\n",
    "            idx_to_remove.append(k)\n",
    "\n",
    "    k += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dd976a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(go_names)>0:\n",
    "    idx_to_remove.reverse()\n",
    "\n",
    "    q_list = q_values.tolist()\n",
    "\n",
    "    for k in idx_to_remove:\n",
    "        go_names.pop(k)\n",
    "        go_tags.pop(k)\n",
    "        q_list.pop()\n",
    "\n",
    "q_values = np.array(q_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "257b5d53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RESPONSE TO CYTOKINE 0.0\n",
      "DEFENSE RESPONSE 0.0\n",
      "RESPONSE TO BIOTIC STIMULUS 4.810186341288886e-267\n",
      "IMMUNE EFFECTOR PROCESS 2.913207773775178e-261\n",
      "REGULATION OF IMMUNE SYSTEM PROCESS 1.036152040884168e-259\n",
      "CELL ACTIVATION 7.81472676526089e-244\n",
      "CYTOKINE PRODUCTION 2.7785081238320015e-191\n",
      "REGULATION OF RESPONSE TO EXTERNAL STIMULUS 5.093393294942207e-185\n",
      "REGULATION OF RESPONSE TO STRESS 2.01706656288609e-181\n",
      "PROTEIN MODIFICATION BY SMALL PROTEIN CONJUGATION OR REMOVAL 4.9355475734853535e-155\n",
      "POSITIVE REGULATION OF SIGNALING 1.7235871316595757e-151\n",
      "POSITIVE REGULATION OF PROTEIN METABOLIC PROCESS 5.414223643482264e-126\n",
      "BIOLOGICAL PROCESS INVOLVED IN SYMBIOTIC INTERACTION 1.1008167292178458e-110\n",
      "POSITIVE REGULATION OF MULTICELLULAR ORGANISMAL PROCESS 4.279581312977233e-106\n",
      "ADAPTIVE IMMUNE RESPONSE 5.92392602017363e-100\n",
      "POSITIVE REGULATION OF GENE EXPRESSION 2.0374405235869985e-99\n",
      "IMMUNE SYSTEM DEVELOPMENT 5.483657377613609e-98\n",
      "REGULATION OF INTRACELLULAR SIGNAL TRANSDUCTION 3.069691271772464e-92\n",
      "MAPK CASCADE 1.721880408017261e-90\n",
      "I KAPPAB KINASE NF KAPPAB SIGNALING 7.632956301512582e-88\n",
      "CELL POPULATION PROLIFERATION 1.3369927272313607e-85\n",
      "SECRETION 2.0175222869423882e-83\n",
      "REGULATION OF PROTEIN MODIFICATION PROCESS 5.651395825712188e-83\n",
      "PATTERN RECOGNITION RECEPTOR SIGNALING PATHWAY 1.058528983926793e-82\n",
      "PROTEIN PHOSPHORYLATION 8.689369154688122e-82\n",
      "APOPTOTIC PROCESS 2.1122299091814854e-81\n",
      "POSITIVE REGULATION OF MOLECULAR FUNCTION 3.3541639299316656e-80\n",
      "CELL KILLING 5.05357288569182e-80\n",
      "REGULATION OF DNA BINDING TRANSCRIPTION FACTOR ACTIVITY 8.732316489372004e-79\n",
      "REGULATION OF CELL DEATH 1.6093420272908492e-78\n",
      "HUMORAL IMMUNE RESPONSE 1.6093420272908492e-78\n",
      "RESPONSE TO OXYGEN CONTAINING COMPOUND 1.453801914216205e-77\n",
      "RECEPTOR SIGNALING PATHWAY VIA STAT 3.60272077892923e-75\n",
      "REGULATION OF PHOSPHORUS METABOLIC PROCESS 7.760820269042518e-74\n",
      "BIOLOGICAL ADHESION 3.181047100035909e-72\n",
      "RESPONSE TO LIPID 4.4724872574728923e-72\n",
      "PROTEOLYSIS 5.989541791720294e-72\n",
      "NEGATIVE REGULATION OF RESPONSE TO STIMULUS 8.202340854663341e-72\n",
      "STRESS ACTIVATED PROTEIN KINASE SIGNALING CASCADE 1.8222067789404604e-71\n",
      "RESPONSE TO NITROGEN COMPOUND 7.401244790731862e-70\n",
      "ANTIGEN PROCESSING AND PRESENTATION 3.692266725127425e-69\n",
      "REGULATION OF TRANSFERASE ACTIVITY 2.3805874789838013e-68\n",
      "MACROMOLECULE CATABOLIC PROCESS 3.369744376189296e-67\n",
      "NEGATIVE REGULATION OF MULTICELLULAR ORGANISMAL PROCESS 3.9291043294013576e-67\n",
      "REGULATION OF CELL DIFFERENTIATION 7.772713358208957e-67\n",
      "PRODUCTION OF MOLECULAR MEDIATOR OF IMMUNE RESPONSE 1.7519252814854489e-65\n",
      "POSITIVE REGULATION OF DEVELOPMENTAL PROCESS 2.1805562203337025e-65\n",
      "PHAGOCYTOSIS 2.275171990378221e-65\n",
      "REGULATION OF MULTICELLULAR ORGANISMAL DEVELOPMENT 7.223928040893083e-65\n",
      "ORGANONITROGEN COMPOUND CATABOLIC PROCESS 1.3555619876052923e-63\n",
      "PEPTIDYL AMINO ACID MODIFICATION 4.470400370014233e-63\n",
      "ENZYME LINKED RECEPTOR PROTEIN SIGNALING PATHWAY 1.7915450351841854e-62\n",
      "NEGATIVE REGULATION OF SIGNALING 2.7985208540209568e-61\n",
      "RESPONSE TO WOUNDING 3.70630502495655e-61\n",
      "REGULATION OF TRANSPORT 5.6498284600552016e-61\n",
      "REGULATION OF CATABOLIC PROCESS 1.0385770774733884e-60\n",
      "NEURON DEATH 1.3123189642826464e-60\n",
      "LOCOMOTION 6.404532557069336e-57\n",
      "COAGULATION 9.134212376537619e-57\n",
      "NIK NF KAPPAB SIGNALING 6.631786921863221e-56\n",
      "RESPONSE TO ENDOGENOUS STIMULUS 1.6879608299708723e-54\n",
      "RESPONSE TO ORGANIC CYCLIC COMPOUND 1.761348136487983e-54\n",
      "TYPE 2 IMMUNE RESPONSE 5.566479526226873e-54\n",
      "REGULATION OF PROTEIN LOCALIZATION 1.4485387070395182e-53\n",
      "MODULATION OF PROCESS OF OTHER ORGANISM 1.6692451122042902e-53\n",
      "REGULATION OF BODY FLUID LEVELS 7.066103321128723e-53\n",
      "POSITIVE REGULATION OF BIOSYNTHETIC PROCESS 1.911021138444521e-52\n",
      "POSITIVE REGULATION OF CELLULAR COMPONENT ORGANIZATION 2.1816269869204093e-52\n",
      "REGULATION OF CELLULAR LOCALIZATION 3.37404728650688e-52\n",
      "REGULATION OF HYDROLASE ACTIVITY 9.544093373148286e-51\n",
      "REGULATION OF CELLULAR COMPONENT MOVEMENT 3.2646140195819065e-50\n",
      "POSITIVE REGULATION OF NUCLEOBASE CONTAINING COMPOUND METABOLIC PROCESS 3.316572710849451e-50\n",
      "HOMEOSTATIC PROCESS 7.303970320418568e-50\n",
      "RESPONSE TO ABIOTIC STIMULUS 8.49752352922106e-50\n",
      "INTRACELLULAR RECEPTOR SIGNALING PATHWAY 1.150971727221084e-49\n",
      "RESPONSE TO INORGANIC SUBSTANCE 2.7533618337899993e-49\n",
      "RESPONSE TO GROWTH FACTOR 2.955360286831491e-48\n",
      "REACTIVE NITROGEN SPECIES METABOLIC PROCESS 2.1876552623229566e-47\n",
      "NEGATIVE REGULATION OF MOLECULAR FUNCTION 2.8759720307899105e-47\n",
      "REGULATION OF BINDING 1.6125157336878038e-46\n",
      "ORGAN OR TISSUE SPECIFIC IMMUNE RESPONSE 2.5040833256034084e-46\n",
      "REACTIVE OXYGEN SPECIES METABOLIC PROCESS 4.758758888771319e-46\n",
      "REGULATION OF ANATOMICAL STRUCTURE MORPHOGENESIS 2.4930429791693793e-45\n",
      "INOSITOL LIPID MEDIATED SIGNALING 4.150848714210413e-45\n",
      "PROCESS UTILIZING AUTOPHAGIC MECHANISM 4.483393413391018e-45\n",
      "NEGATIVE REGULATION OF DEVELOPMENTAL PROCESS 9.207123702322071e-45\n",
      "CELLULAR RESPONSE TO EXTERNAL STIMULUS 1.8273209298567638e-44\n",
      "CIRCULATORY SYSTEM DEVELOPMENT 2.0691391671983214e-44\n",
      "GLAND DEVELOPMENT 4.0405855115380444e-44\n",
      "ANATOMICAL STRUCTURE FORMATION INVOLVED IN MORPHOGENESIS 6.979759293053695e-44\n",
      "ENDOCYTOSIS 7.66010904775091e-44\n",
      "TUBE DEVELOPMENT 1.3757952886443496e-43\n",
      "CELLULAR RESPONSE TO DNA DAMAGE STIMULUS 1.766117207554216e-43\n",
      "GROWTH 4.832999309580766e-43\n",
      "NECROTIC CELL DEATH 6.227537518483309e-43\n",
      "PRI MIRNA TRANSCRIPTION BY RNA POLYMERASE II 9.620764791306591e-43\n",
      "REGULATION OF DNA TEMPLATED TRANSCRIPTION IN RESPONSE TO STRESS 1.7791046801247676e-42\n",
      "PROTEIN KINASE B SIGNALING 3.083018107418663e-42\n",
      "MAINTENANCE OF LOCATION 3.558592374433875e-42\n",
      "REGULATION OF CELLULAR COMPONENT BIOGENESIS 3.54752749903366e-41\n",
      "TISSUE REMODELING 4.5171373236527903e-41\n",
      "CELLULAR RESPONSE TO CHEMICAL STRESS 9.504233787078641e-41\n",
      "AGING 9.623724187618086e-41\n",
      "NEGATIVE REGULATION OF GENE EXPRESSION 1.2569703311414152e-40\n",
      "SECOND MESSENGER MEDIATED SIGNALING 6.204308614817647e-39\n",
      "RESPONSE TO ENDOPLASMIC RETICULUM STRESS 6.918805819705277e-39\n",
      "RESPONSE TO OXIDATIVE STRESS 1.7495588798210798e-38\n",
      "INTEGRIN MEDIATED SIGNALING PATHWAY 2.5582487767463493e-38\n",
      "NITRIC OXIDE SYNTHASE BIOSYNTHETIC PROCESS 1.015950204582225e-37\n",
      "CELL MORPHOGENESIS 2.199240723756248e-37\n",
      "POST TRANSLATIONAL PROTEIN MODIFICATION 2.259004784082319e-37\n",
      "REGULATION OF OXIDOREDUCTASE ACTIVITY 3.0718007518497683e-37\n",
      "RECEPTOR METABOLIC PROCESS 3.856857230146972e-37\n",
      "REGULATION OF ORGANELLE ORGANIZATION 1.2835588830831308e-36\n",
      "NEGATIVE REGULATION OF PROTEIN METABOLIC PROCESS 1.5303982578681323e-36\n",
      "ESTABLISHMENT OF PROTEIN LOCALIZATION 2.2005653597393153e-36\n",
      "NEUROGENESIS 2.407918096655603e-36\n",
      "REGULATION OF PROTEIN STABILITY 2.9518101008452002e-36\n",
      "CELL CELL SIGNALING 5.507495248148456e-36\n",
      "LYSOSOME LOCALIZATION 7.091278628800628e-36\n",
      "MEMBRANE ORGANIZATION 1.523704817278872e-35\n",
      "SMALL GTPASE MEDIATED SIGNAL TRANSDUCTION 4.5365351872468994e-35\n",
      "RESPONSE TO DRUG 6.644063066537683e-35\n",
      "PROGRAMMED CELL DEATH INVOLVED IN CELL DEVELOPMENT 8.285047878910216e-35\n",
      "POSITIVE REGULATION OF SMALL MOLECULE METABOLIC PROCESS 2.784631978253178e-34\n",
      "VESICLE ORGANIZATION 3.120808639073862e-34\n",
      "PYROPTOSIS 3.640340767099224e-34\n",
      "FOAM CELL DIFFERENTIATION 3.888841330227644e-34\n",
      "CYTOLYSIS 5.754292041209309e-34\n",
      "ESTABLISHMENT OF ORGANELLE LOCALIZATION 7.956825953436298e-34\n",
      "RESPONSE TO TOPOLOGICALLY INCORRECT PROTEIN 1.9228780984348566e-33\n",
      "HORMONE TRANSPORT 4.772489467953026e-33\n",
      "PROTEIN CONTAINING COMPLEX SUBUNIT ORGANIZATION 5.284224180559323e-33\n",
      "PROTEIN O LINKED GLYCOSYLATION 7.60279292732753e-33\n",
      "EMBRYO IMPLANTATION 7.60279292732753e-33\n",
      "RESPONSE TO ANTIBIOTIC 1.5127823490053672e-32\n",
      "OSSIFICATION 1.8597183972454289e-32\n",
      "PROTEIN DEPHOSPHORYLATION 3.5406879905906465e-32\n",
      "INTRACELLULAR TRANSPORT 2.2233307605705653e-31\n",
      "CELLULAR MACROMOLECULE LOCALIZATION 3.0152965017713305e-31\n",
      "PROTEIN FOLDING IN ENDOPLASMIC RETICULUM 5.393172730331434e-31\n",
      "RESPONSE TO ARSENIC CONTAINING SUBSTANCE 1.0489644951070607e-30\n",
      "PROTEIN REFOLDING 1.0967089374400363e-30\n",
      "AMYLOID BETA CLEARANCE 1.2657596538591545e-30\n",
      "MHC CLASS II BIOSYNTHETIC PROCESS 1.3257039728309245e-30\n",
      "ACTIN FILAMENT BASED PROCESS 1.3502361514904686e-30\n",
      "ENDOTHELIUM DEVELOPMENT 1.4323285485811008e-30\n",
      "CALCIUM ION TRANSPORT 3.282748954963372e-30\n",
      "POSITIVE REGULATION OF CARBOHYDRATE METABOLIC PROCESS 3.3097423932793935e-30\n",
      "REGULATION OF ANATOMICAL STRUCTURE SIZE 6.897232743855471e-30\n",
      "RESPONSE TO NICOTINE 6.897232743855471e-30\n",
      "ER NUCLEUS SIGNALING PATHWAY 8.455718291621137e-30\n",
      "LIPID PHOSPHORYLATION 9.078654159310398e-30\n",
      "ESTABLISHMENT OF CELL POLARITY 9.972553777705693e-30\n",
      "MORPHOGENESIS OF AN EPITHELIUM 9.984708298985504e-30\n",
      "CARBOHYDRATE TRANSMEMBRANE TRANSPORT 1.0077298811848324e-29\n",
      "FAT CELL DIFFERENTIATION 1.0877627599802312e-29\n",
      "REGULATION OF CELLULAR RESPIRATION 1.48348594809351e-29\n",
      "DNA METABOLIC PROCESS 2.5883410855774423e-29\n",
      "TISSUE MIGRATION 3.273196671747016e-29\n",
      "HYALURONAN METABOLIC PROCESS 3.849637760949928e-29\n",
      "CELL PROJECTION ORGANIZATION 4.2422392031017975e-29\n",
      "REGULATION OF SYNAPSE STRUCTURE OR ACTIVITY 4.938007049044053e-29\n",
      "MULTI ORGANISM CELLULAR PROCESS 5.926203184780019e-29\n",
      "EXTERNAL ENCAPSULATING STRUCTURE ORGANIZATION 6.039896962190394e-29\n",
      "ENDOCRINE SYSTEM DEVELOPMENT 6.61409256004972e-29\n",
      "RHYTHMIC PROCESS 7.200279407512464e-29\n",
      "RESPONSE TO ISCHEMIA 1.5876797962835424e-28\n",
      "REGULATION OF LIPID METABOLIC PROCESS 2.315685862669923e-28\n",
      "LONG CHAIN FATTY ACID IMPORT INTO CELL 6.319293846601313e-28\n",
      "CELLULAR COMPONENT MORPHOGENESIS 6.429409489862031e-28\n",
      "REGULATION OF MITOCHONDRIAL MEMBRANE POTENTIAL 9.205413173652598e-28\n",
      "AMYLOID FIBRIL FORMATION 1.2042384718595119e-27\n"
     ]
    }
   ],
   "source": [
    "for a,b in zip(go_names,q_values):\n",
    "    print(a,b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af277f53",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
